### ISSUE TITLE
Eigen-Tune: feature advertised as functional but pipeline is incomplete and not wired into the runtime

### ISSUE THREAD

**Issue by jameslescroc** [temm1e-labs/temm1e#35]
## Summary

The `temm1e-distill` crate (Eigen-Tune: Self-Tuning Knowledge Distillation Engine) is documented and advertised as a working feature (setup guide, `[eigentune]` config section, `/eigentune` slash commands, automatic training/evaluation/graduation), but in its current state:

1. Roughly half of the pipeline described in `tems_lab/eigen/IMPLEMENTATION.md` is not implemented.
2. The crate is not consumed by any other crate in the workspace — no runtime code ever instantiates `EigenTuneEngine`.
3. Following the public setup guide produces no observable effect: no data is collected, no model is trained, no slash command is recognized.

The statistical core, the SQLite store, the state machine and the collector are well written and tested in isolation, so this is not a rewrite — it is a "finish the wiring + fill the missing modules" issue.

## Environment

- Repo: `temm1e-main` at current `main`
- Crate: `crates/temm1e-distill` (~7 900 LOC, 27 files, 156 test annotations)
- Rust edition 2021, MSRV 1.82
- Verified by static analysis only (toolchain not available in review environment). `cargo check -p temm1e-distill` and the two integration tests should still be run to confirm the isolated parts compile and pass.

## What works today

The following is implemented, documented and covered by tests:

- `src/collector.rs` — fire-and-forget capture of `(request, response)` pairs with domain classification and LRU-style eviction per tier.
- `src/scorer.rs` — Beta-Binomial quality scoring.
- `src/store.rs` (~1 136 LOC) — full SQLite schema, CRUD, tier records, training runs table.
- `src/config.rs` (~447 LOC) — `EigenTuneConfig` with validation and TOML defaults.
- `src/stats/` — SPRT (Wald 1945), CUSUM (Page 1954), Wilson score intervals, Shannon entropy, Thompson sampling, Beta utilities, power analysis. All pure, all unit-tested.
- `src/engine/state_machine.rs` — per-tier state machine `Collecting → Training → Evaluating → Shadowing → Graduated` with transitions guarded by the statistical tests above.
- `src/engine/router.rs`, `shadow.rs`, `monitor.rs`, `graduation.rs` — routing decision, SPRT shadow coordinator, CUSUM drift monitor, promotion/demotion manager.
- `src/backends/ollama.rs` — health check, model listing, model creation from a GGUF path, embedding generation.
- `src/judge/embedding.rs`, `src/judge/behavior.rs` — cosine similarity judge and user-behavior judge.
- `tests/proof_of_pipeline.rs` (687 LOC) and `tests/bench_eigentune.rs` (1 405 LOC) — end-to-end tests against an in-memory SQLite.

Zero `todo!` / `unimplemented!` / `panic!` / `FIXME` / `XXX` in `crates/temm1e-distill`.

## What is missing versus `tems_lab/eigen/IMPLEMENTATION.md`

The design document specifies the following files that do not exist in `src/`:

| Planned file | Status | Impact |
|---|---|---|
| `src/curator.rs` | **Missing** | Dataset building, dedup, balancing and ChatML JSONL export are re-implemented inline inside `tests/proof_of_pipeline.rs`. No reusable API. |
| `src/status.rs` | **Missing** | Status logic lives inline in `lib.rs::format_status`. Minor. |
| `src/engine/trainer.rs` | **Missing — critical** | No orchestrator spawns Unsloth/MLX. `state_machine.rs:33` has `TierState::Training => Ok(None), // Training transitions handled by trainer`. A tier entering `Training` is stuck there forever. |
| `src/engine/evaluator.rs` | **Missing — critical** | No benchmark runner writes `eval_accuracy` / `eval_n` on `tier_records`. The `Evaluating → Shadowing` transition in `state_machine.rs:83-118` reads fields that are never written in production. |
| `src/backends/unsloth.rs` | **Missing** | Setup guide lists Unsloth as a supported training backend. No Rust code launches it. |
| `src/backends/mlx.rs` | **Missing** | Same for Apple Silicon. |
| `src/backends/hf_autotrain.rs` | **Missing** | Advertised in `IMPLEMENTATION.md`. |
| `src/judge/teacher.rs` | **Missing** | Opt-in LLM-as-judge premium mode. Lowest priority. |

The only backend that exists is `ollama.rs`, which is inference-only. The prerequisite check in `lib.rs:230-268` runs `python3 -c "import mlx_lm"` and `import unsloth` via `std::process::Command` **only to render a ✓/✗ in the status view** — no code anywhere actually launches a fine-tuning run.

### 1. The pipeline never closes the loop

```
Collecting  ──▶  Training  ──▶  (dead end)
                                │
                                ▼
                    Evaluating / Shadowing / Graduated
                    reachable only by writing to SQLite by hand
```

Because `trainer.rs` and `evaluator.rs` do not exist, no tier ever leaves `Training`. Consequently no tier ever reaches `Shadowing` or `Graduated` through normal operation, and `EigenTuneRouter::route()` always returns `RouteDecision::Cloud`.

### 2. The crate is not wired into the runtime

Search results for `temm1e-distill` / `temm1e_distill` / `EigenTune` outside the crate itself:

- `Cargo.toml:22` — declared as a workspace member
- `Cargo.toml:143` — declared as a workspace dependency
- `crates/temm1e-perpetuum/src/conscience.rs:163` — a **comment only**: `// Dream completes externally (EigenTune signals done)`

No `use temm1e_distill::...` anywhere. Specifically:

- `src/main.rs` (binary entry) does not depend on `temm1e-distill` and does not expose an `eigentune` subcommand.
- `crates/temm1e-agent` does not call `EigenTuneEngine::on_completion`, `on_signal`, `route`, `tick`, `on_shadow_observation`, or `on_monitor_observation`.
- `crates/temm1e-gateway` does not instantiate the engine.
- No slash command parser recognizes `/eigentune status`, `/eigentune model`, `/eigentune setup`.

### 3. The public setup guide does not match reality

The user-facing documentation (setup guide) says:

> Add to your `temm1e.toml`:
>
> ```toml
> [eigentune]
> enabled = true
> ```
>
> That's it. Restart TEMM1E and the system begins collecting training data from every conversation.

Actual behavior after following this procedure:

- No rows are inserted into `eigentune_pairs`.
- `/eigentune status` returns "unknown command".
- No background task runs `GraduationManager::tick()`.
- Prerequisites check (`is_available()` for Ollama, `import mlx_lm`) is never executed because `check_prerequisites()` is never called.

The guide is effectively a design preview, not a user guide for the current code.

## Reproduction

1. `cargo build --release --bin temm1e`
2. Add `[eigentune]\nenabled = true` to `temm1e.toml`.
3. Start: `./target/release/temm1e start` (or `chat` / `tui`).
4. Have a few conversations.
5. Expected per the setup guide: training pairs accumulate, `/eigentune status` shows counts, training triggers at 500+ pairs.
6. Actual: `sqlite3 <db> "SELECT COUNT(*) FROM eigentune_pairs;"` returns `0`. `/eigentune status` is not a recognized command. No `eigentune_*` tracing spans appear in logs.

Additionally, to confirm the isolated parts still work:

```bash
cargo check -p temm1e-distill
cargo clippy -p temm1e-distill --all-targets --all-features -- -D warnings
cargo test -p temm1e-distill
cargo test -p temm1e-distill --test proof_of_pipeline -- --nocapture
cargo test -p temm1e-distill --test bench_eigentune -- --nocapture
```

These are expected to pass — the issue is about integration and missing modules, not about the existing code being broken.

## Proposed resolution

Two options. Option A is honest and fast; option B is the real fix.

### Option A — Align documentation with reality (short term)

- Move `docs/eigentune/SETUP.md` (or equivalent) under `tems_lab/eigen/` next to `DESIGN.md` and `IMPLEMENTATION.md`.
- Add a banner at the top: `> Status: design preview. The runtime does not yet execute this pipeline. Tracking issue: #<this>`.
- Remove the `/eigentune` commands from any user-visible help text until they exist.
- Add a `#![doc = "Status: experimental, not wired into runtime"]` warning on `crates/temm1e-distill/src/lib.rs`.

Effort: ~1 hour.

### Option B — Finish the implementation (real fix)

Ordered by dependency:

1. **`src/engine/trainer.rs`** — async orchestrator that:
   - Reads high-quality pairs for a tier via `EigenTuneStore`.
   - Writes a ChatML JSONL file to a temp path.
   - Spawns `mlx_lm.lora` (Apple Silicon) or `unsloth` (NVIDIA) via `tokio::process::Command`, streams stdout/stderr to `tracing`.
   - Produces a GGUF file, calls `backends::ollama::create_model`.
   - Writes a `TrainingRun` row, updates the tier state to `Evaluating`.
   - On failure: tier returns to `Collecting`, error is logged, never propagated.

2. **`src/engine/evaluator.rs`** — runs the holdout set against the newly created Ollama model, compares with `judge/embedding.rs`, writes `eval_accuracy` and `eval_n` on the `tier_records` row, lets `state_machine::check_evaluating_transition` do the rest.

3. **`src/curator.rs`** — extract the dataset-building logic from `tests/proof_of_pipeline.rs` into a reusable module (dedup by hash, Thompson sampling for balance, ChatML export).

4. **Backends**:
   - `src/backends/mlx.rs` — `fn train(dataset_path, base_model, output_dir) -> Result<PathBuf>` that wraps the `mlx_lm.lora` CLI.
   - `src/backends/unsloth.rs` — same interface, wraps a short Python driver script shipped as `scripts/eigentune_unsloth.py`.
   - Dispatch in `src/backends/mod.rs` based on `EigenTuneConfig::backend`.

5. **Runtime wiring**:
   - Add `temm1e-distill = { workspace = true }` to `crates/temm1e-agent/Cargo.toml`.
   - In the agent constructor, build `EigenTuneEngine` if `config.eigentune.enabled`.
   - In the agent loop: call `route()` before `Provider.complete()` to pick Cloud vs Local; call `on_completion()` after every completion; call `on_signal()` on user reactions (retry, thumbs, continuation); spawn a periodic `tick()` task every N seconds.
   - Add a `graceful_shutdown` hook that flushes the collector.

6. **CLI**:
   - In `src/main.rs`, add a clap subcommand `eigentune` with:
     - `status` → prints `EigenTuneEngine::format_status()`.
     - `setup` → prints the prerequisite check + install hints.
     - `model [name|auto]` → reads/writes `base_model`.
     - `tick` → forces one graduation cycle.
   - Register the same as chat slash commands in `temm1e-channels`.

7. **Tests**:
   - Add a feature-gated integration test `tests/trainer_smoke.rs` that runs only when `MLX_AVAILABLE=1` or `UNSLOTH_AVAILABLE=1`, trains a tiny model on 50 fake pairs, verifies the GGUF is registered in Ollama.
   - Add a unit test per new backend for the command construction (no process spawn).

8. **Docs**:
   - Update the setup guide to reference the actual commands that now exist.
   - Add a troubleshooting section pointing at `tracing` targets (`temm1e_distill::collector`, `::engine::trainer`, etc.).

## Acceptance criteria

For this issue to be closed, the following must be true on a fresh clone:

- [ ] `cargo check --workspace` passes.
- [ ] `cargo clippy --workspace --all-targets --all-features -- -D warnings` passes.
- [ ] `cargo test -p temm1e-distill` passes (including existing `proof_of_pipeline` and `bench_eigentune`).
- [ ] With `[eigentune] enabled = true` in `temm1e.toml`, starting the binary and having at least one conversation inserts a row into `eigentune_pairs`.
- [ ] `temm1e eigentune status` prints a non-empty report including prerequisites, pair counts, per-tier state, and diversity J.
- [ ] With MLX or Unsloth installed and `ollama serve` running, a tier that reaches `min_pairs` with sufficient diversity actually transitions `Collecting → Training → Evaluating → Shadowing → Graduated` without manual SQL intervention.
- [ ] The public setup guide matches the actually shipped commands and behaviors.
- [ ] `CLAUDE.md` no longer lists `temm1e-distill` as a runtime component unless it is truly wired in.

## Related files and references

- `crates/temm1e-distill/src/lib.rs` — public API, `EigenTuneEngine`, `check_prerequisites`, `format_status`.
- `crates/temm1e-distill/src/engine/state_machine.rs:33` — the `Training` dead end.
- `crates/temm1e-distill/src/engine/state_machine.rs:83-118` — `Evaluating → Shadowing` that depends on fields never written in production.
- `crates/temm1e-distill/src/backends/ollama.rs` — the only existing backend.
- `crates/temm1e-distill/tests/proof_of_pipeline.rs` — contains the curator logic that should move to `src/curator.rs`.
- `tems_lab/eigen/IMPLEMENTATION.md` — original build plan.
- `tems_lab/eigen/DESIGN.md` — architectural rationale.
- `Cargo.toml:22,143` — workspace registration (only external reference to the crate).
- `crates/temm1e-perpetuum/src/conscience.rs:163` — stale comment referencing EigenTune.

**Comment by nagisanzenin** [comment]
## Fixed in v4.9.0

PR #36 merged — every issue documented here is now resolved:

**Missing modules (§"What is missing"):**
- ✅ `src/curator.rs` — dataset building (dedup, balance via Thompson sampling, stratified holdout split, ChatML JSONL export)
- ✅ `src/engine/trainer.rs` — orchestrator that runs curator → backend → Ollama commit, with wall-clock timeout + failure recovery
- ✅ `src/engine/evaluator.rs` — runs eval holdout against trained Ollama model, writes `eval_accuracy` + `eval_n`
- ✅ `src/backends/mlx.rs` — Apple Silicon MLX trainer subprocess wrapper
- ✅ `src/backends/unsloth.rs` — NVIDIA Unsloth Python wrapper invoker + vendored `scripts/eigentune_unsloth.py`
- ✅ `src/backends/mod.rs` — `TrainingBackend` trait + `select_backend()` dispatch

**State machine dead end (§"What is functionally broken"):**
- ✅ `state_machine.rs:33` `TierState::Training => Ok(None)` replaced with `check_training_transition()` that recovers stuck Training tiers after 1 hour

**Not wired into runtime (§"The crate is not wired into the runtime"):**
- ✅ `crates/temm1e-agent/src/runtime.rs` gains `eigen_tune` field + `with_eigen_tune` builder + 5 fire-and-forget hooks (complexity capture, full routing wrapper, collection hook, tool result signal, user message signal)
- ✅ `src/main.rs` constructs the engine via two-pass TOML parse, injects it into the agent, spawns a 60s periodic tick task — wired into BOTH `Commands::Start` (daemon) and `Commands::Chat` (CLI) paths
- ✅ CLI subcommand: `temm1e eigentune {status, setup, model, tick, demote}`
- ✅ Slash command: `/eigentune ...` in BOTH gateway and CLI parsers

**Setup guide mismatch (§"The public setup guide does not match reality"):**
- ✅ `tems_lab/eigen/SETUP.md` completely replaced with the real protocol: double opt-in (`enabled` + `enable_local_routing`), compatibility matrix, seven-gate safety chain, troubleshooting

**Documentation overclaims:**
- ✅ `README.md` v3.1.0 changelog corrected ("proven on M2" was a manual one-off; runtime integration deferred to v4.9.0)
- ✅ `README.md` v4.9.0 entry documents the real feature set
- ✅ `README.md` Eigen-Tune section rewritten with double opt-in and safety chain
- ✅ `CLAUDE.md` clarifies runtime gating

**Acceptance criteria:**
- ✅ `cargo check --workspace` passes
- ✅ `cargo clippy --workspace --all-targets -- -D warnings` passes
- ✅ `cargo test -p temm1e-distill` passes (existing `proof_of_pipeline` and `bench_eigentune` + 30 new tests)
- ✅ With `[eigentune] enabled = true`, starting the binary and having conversations inserts rows into `eigentune_pairs` (verified: 10 pairs collected from 10-turn live test)
- ✅ `temm1e eigentune status` prints a non-empty report including prerequisites, pair counts, per-tier state, and both opt-in switches
- ✅ The public setup guide matches the actually shipped commands and behaviors

**What was NOT done (deferred to follow-up PRs):**
- The full Collecting → Training → Evaluating → Shadowing → Graduated lifecycle with a REAL fine-tuning run was not performed end-to-end (the machine's default `python3` is 3.9.6; MLX needs 3.10+). The Rust code that drives the trainer is complete and unit-tested; the actual `python3 -m mlx_lm.lora` subprocess invocation awaits a host with Python 3.12+ as the default `python3`.
- General-purpose data mixing (catastrophic forgetting prevention)
- GGUF conversion for non-ADAPTER model families (restricted to Llama/Mistral/Gemma for MVP)
- Teacher judge (LLM-as-judge)
- Tool-use-capable local model training

**New documentation produced:**
- `tems_lab/eigen/INTEGRATION_PLAN.md` (1,932 lines) — 22-phase master plan
- `tems_lab/eigen/CODE_ANCHORS.md` (1,054 lines) — verified file:line citations
- `tems_lab/eigen/LOCAL_ROUTING_SAFETY.md` (540 lines) — seven-gate safety chain spec

**Test count:** 2,337 workspace tests passing, 0 failing. Default-config users see zero behavior change.

**Comment by nagisanzenin** [comment]
## Follow-up: Real end-to-end fine-tuning now proven (PR #37)

The deferred item from the v4.9.0 close-out is now complete:

> The full Collecting → Training → Evaluating → Shadowing → Graduated lifecycle with a REAL fine-tuning run was not performed end-to-end

**What we did in #37:**

1. **Set up Python 3.13 venv** with `mlx-lm` on Apple M2 16GB
2. **MLX LoRA fine-tuned** `mlx-community/Llama-3.2-1B-Instruct-4bit` on 20 ChatML pairs (Rust Q&A), 20 iterations
   - Val loss: 5.394 → **1.387** (73% reduction)
   - Peak memory: 1.59 GB, trainable params: 5.6M/1.2B (0.46% LoRA)
3. **Fused adapters** → dequantized to f16 → converted to GGUF via llama.cpp `convert_hf_to_gguf.py` → quantized to Q4_K_M (807 MB)
4. **Loaded into Ollama** and verified inference through OpenAI-compatible API
5. **Routed through AgentRuntime** with EigenTune wired in — tier set to Graduated → router returned `Local(endpoint)` → request went to `localhost:11434` → **cloud provider was never called**

**Bugs found & fixed:**
- `resolve_python()` needed a 5s timeout per candidate (broken venvs can hang indefinitely)
- `mlx_lm`'s native GGUF export (`--export-gguf`) produces files that crash Ollama — must use llama.cpp's `convert_hf_to_gguf.py` instead
- Ollama's `--experimental` safetensors import with `--quantize` requires MLX dynamic library that Homebrew Ollama doesn't bundle
- The `commit_to_ollama()` ADAPTER directive doesn't work with MLX adapters — must fuse + dequantize + convert to GGUF

**New tests:** 7 mock routing tests + 2 live Ollama integration tests (`EIGENTUNE_LIVE=1`). Total: 2,346 tests, 0 failures.
